altair/vegalite/v5/api.py

```diff
@@ -380,24 +422,625 @@ def check_fields_and_encodings(parameter: Parameter, field_name: str) -> bool:
+        - Due to the assumption that other `UndefinedType`'s could exist.
+    - Current [typing spec advises](https://typing.readthedocs.io/en/latest/spec/concepts.html#support-for-singleton-types-in-unions) using an `Enum`.
+        - Otherwise, requires an explicit guard to inform the type checker.
+    """
+    return obj is Undefined
+
+
+def _get_predicate_expr(p: Parameter) -> Optional[str | SchemaBase]:
+    # https://vega.github.io/vega-lite/docs/predicate.html
+    return getattr(p.param, "expr", Undefined)
+
+
+def _predicate_to_condition(
+    predicate: _PredicateType, *, empty: bool | UndefinedType = Undefined
+) -> _ConditionType:
+    condition: _ConditionType
+    if isinstance(predicate, Parameter):
+        predicate_expr = _get_predicate_expr(predicate)
+        if predicate.param_type == "selection" or _is_undefined(predicate_expr):
+            condition = {"param": predicate.name}
+            if isinstance(empty, bool):
+                condition["empty"] = empty
+            elif isinstance(predicate.empty, bool):
+                condition["empty"] = predicate.empty
+        else:
+            condition = {"test": predicate_expr}
+    elif _is_test_predicate(predicate):
+        condition = {"test": predicate}
+    elif isinstance(predicate, dict):
+        condition = predicate
+    elif isinstance(predicate, _expr_core.OperatorMixin):
+        condition = {"test": predicate._to_expr()}
+    else:
+        msg = f"condition predicate of type {type(predicate).__name__!r}"
+        raise NotImplementedError(msg)
+    return condition
+
+
+def _condition_to_selection(
+    condition: _ConditionType,
+    if_true: _StatementType,
+    if_false: _StatementType,
+    **kwargs: Any,
+) -> _SelectionType:
+    selection: _SelectionType
+    if isinstance(if_true, core.SchemaBase):
+        if_true = if_true.to_dict()
+    elif isinstance(if_true, str):
+        if isinstance(if_false, str):
+            msg = (
+                "A field cannot be used for both the `if_true` and `if_false` "
+                "values of a condition. "
+                "One of them has to specify a `value` or `datum` definition."
+            )
+            raise ValueError(msg)
+        else:
+            if_true = utils.parse_shorthand(if_true)
+            if_true.update(kwargs)
+    condition.update(if_true)
+    if isinstance(if_false, core.SchemaBase):
+        # For the selection, the channel definitions all allow selections
+        # already. So use this SchemaBase wrapper if possible.
+        selection = if_false.copy()
+        selection.condition = condition
+    elif isinstance(if_false, (str, dict)):
+        if isinstance(if_false, str):
+            if_false = utils.parse_shorthand(if_false)
+            if_false.update(kwargs)
+        selection = dict(condition=condition, **if_false)
+    else:
+        raise TypeError(if_false)
+    return selection
+
+
+class _Conditions(TypedDict, total=False):
```

Comment:
This is helpful for making sense of the internal structure. 

An issue it poses can be seen in `alt.Chart` during [this test](https://github.com/dangotbanned/altair/blob/25ac2fa4601a02c340ba20be1d6dc858e3e18aad/tests/vegalite/v5/test_api.py#L375-L376)

Using a `TypedDict` makes `mypy` unhappy, as it recognizes `dict` as an accepted type - but not the kind produced by the `TypedDict` metaclass [alt.Chart.encode](https://github.com/vega/altair/blob/7f05ecbf999766ae1398dcbeb6144c128413026c/altair/vegalite/v5/schema/channels.py#L34206-L34216)

Reply:
Fixed in https://github.com/vega/altair/commit/91a1de17f97d0b1de20724f3cfa771f07741dcc7